### PR TITLE
Replace phpunit/token-stream with nikic/php-parser

### LINKED ISSUE #2282: Get rid of phpunit/php-token-stream
We should replace `phpunit/php-token-stream` with something else because it is abandoned.

***
 
https://github.com/contao/contao/pull/2281#issuecomment-690348360 @leofeyer Can we maybe rework the class so it works without the package? Maybe using the PHP function `token_get_all()`?

https://github.com/contao/contao/pull/2281#issuecomment-690352583 @m-vo That's what `PHP_Token_Stream` is doing as well: https://github.com/sebastianbergmann/php-token-stream/blob/master/src/Stream.php#L413-L514

Maybe this function gets small enough if we reduce it to our use case to adopt it.

I had a quick look at it. Parsing and stripping PHP tags is the easy part. Unfortunately we would have to write/adapt quite a lot of code to handle all the namespace edge cases, which makes it - well - a doable but non-trivial task.

https://github.com/contao/contao/pull/2281#issuecomment-690907506 @leofeyer Maybe we can use https://github.com/nikic/PHP-Parser?

### PR DESCRIPTION
**Pull request by m-vo**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2282 
| Docs PR or issue | -

Here is a POC to replace `phpunit/token-stream` with `nikic/php-parser`.

It won't be able to keep the whitespaces in the output, though (see test output, [Reference from docs](https://github.com/nikic/PHP-Parser/blob/master/doc/0_Introduction.markdown#what-output-does-it-produce)). Not sure if this matters as internally we only use this to dump/combine code, not to format it.

### CODE REVIEW

File: core-bundle/src/Config/Loader/PhpFileLoader.php [hunk 1]
```diff
@@ -44,58 +50,63 @@ public function supports($resource, $type = null): bool
...
-                        $code .= $token;
-                    } else {
-                        $namespace = $token->getName();
-                        $stream->seek($token->getEndTokenId());
+        // Parse input into an AST.
+        $contents = trim(file_get_contents($file));
+        $parser = (new ParserFactory())->create(ParserFactory::PREFER_PHP7);
+        $ast = $parser->parse($contents);
+
+        // Traverse AST to handle namespacing and drop "declare('strict_types')"
```
Comment (leofeyer) [inline core-bundle/src/Config/Loader/PhpFileLoader.php]: Should we not keep the first `declare(strict_types=1);`?

Comment (leofeyer) [reply core-bundle/src/Config/Loader/PhpFileLoader.php]: Ah, we are already removing them now, so this is fine. 👍🏻

**Comment by leofeyer** [comment]
> It won't be able to keep the whitespaces in the output, though (see test output

Where can I find the test output?

**Comment by m-vo** [comment]
Easiest to see if you run the tests inside `PhpFileLoaderTest.php` in PHPStorm and look at at the diffs from the failed asserts (`Failed asserting that two strings are identical.`).

**Comment by leofeyer** [comment]
> `$GLOBALS['TL_TEST'] = \true;`

What? Why did the script add the backslash?

**Comment by m-vo** [comment]
It's caused by the `NameResolver` that turns every statement into a fully qualified one which then is [printed like this](https://github.com/nikic/PHP-Parser/blob/88e519766fc58bd46b8265561fb79b54e2e00b28/lib/PhpParser/PrettyPrinter/Standard.php#L61-L63). Strange this also affects `true`. :thinking: 

Edit: Probably because `true` is treated as a global constant. See https://github.com/nikic/PHP-Parser/blob/master/doc/component/Name_resolution.markdown#name-resolution

**Comment by ausi** [comment]
> Probably because `true` is treated as a global constant.

I think `true` **is** a constant. Before PHP 8 you could even redefine it 😳 https://3v4l.org/ng9dU

**Comment by m-vo** [comment]
> I think `true` **is** a constant. Before PHP 8 you could even redefine it  https://3v4l.org/ng9dU

Wait, what?! I didn't now that :scream:.

Edit: But you cannot redefine the lowercase version. This is weird.

File: core-bundle/src/Config/Loader/PhpFileLoader.php [hunk 2]
```diff
@@ -44,71 +48,82 @@ public function supports($resource, $type = null): bool
...
+                // Drop legacy access check ("if(*) die('You cannot access this file directly!');").
+                if ($node instanceof Node\Stmt\If_) {
+                    foreach ($node->stmts as $statement) {
+                        if (
+                            $statement instanceof Node\Stmt\Expression &&
+                            $statement->expr instanceof Node\Expr\Exit_ &&
+                            $statement->expr->expr instanceof Node\Scalar\String_ &&
+                            $this->matchLegacyCheckText($statement->expr->expr->value)
+                        ) {
+                            return NodeTraverser::REMOVE_NODE;
```
Comment (ausi) [inline core-bundle/src/Config/Loader/PhpFileLoader.php]: Wouldn’t this remove all `if` statements that have `die('You cannot access this file directly!')` as one of their sub-statements?

I think we should check for `TL_ROOT` too and make sure that the `die()` statement is the first one.

Comment (m-vo) [reply core-bundle/src/Config/Loader/PhpFileLoader.php]: Checking against all parts in 917fb7548a317443010c87e74fdcddc5ec03338b

**Comment by m-vo** [comment]
50584e20761bf2a4566448e004f319b1010d4a84 concludes the "code style" changes in the reconstructed output. 

This includes:
- whitespaces
- linebreaks
- comments
- prefixed `\`

With this, tests are green again. We should carefully review the namespace implications though and maybe add more tests.

**Comment by m-vo** [comment]
Btw.: we could in theory not apply the namespace transformations (and therefore won't get things like `\true`) with the following:
```diff
- $namespaceResolver = new NameResolver();
+ $namespaceResolver = new NameResolver(null, ['replaceNodes' => false]);
```
But then stripping namespace + use declarations is much more complicated again. (I think it's correct to do the transformations anyway.)

**Comment by ausi** [comment]
> Btw.: we could in theory not apply the namespace transformations

I don’t think that is necessary. AFAIK `\true` is perfectly fine and doesn’t cause any problems, right? The code is for caching, so the readability should not be a concern IMO.

**Comment by leofeyer** [comment]
Ok, so in my tests with the official demo, this worked well. However, can we be sure that it does not have side-effects?

**Comment by m-vo** [comment]
> However, can we be sure that it does not have side-effects?

It's a bit like with every change isn't it? The more comprehensive the tests, the safer it gets. If someone's got some exotic scenarios, sample code would be appreciated (or at least running this PR).

**Comment by ausi** [comment]
Is this actually a bugfix? Or would it be better to target 4.11?

**Comment by m-vo** [comment]
I think both would be valid. `phpunit/token-stream` is abandoned, so this would 'fix' this usage.

**Comment by ausi** [comment]
> `phpunit/token-stream` is abandoned

But it is compatible with PHP 8 and therefore should be sufficient for the lifetime of Contao 4.9 I think.

**Comment by leofeyer** [comment]
@m-vo Can you please rebase the PR onto the master branch then?

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**